### DISCUSSION TITLE
How does progress work?

### DISCUSSION THREAD

**Discussion by CoconutMacaroon** [Eugeny/tabby#4764] [Q&A] [answered]
I recently installed `tabby`, and it works really well. But I'm curious, how does it figure out the current progress for a command? If anyone knows where in the source that is, I can read that too.

Thanks!

**Comment by Eugeny** [comment, marked as answer]
It's scanning for percentages in shell output - check out baseTerminalTab.component.ts

**Reply by CoconutMacaroon** [reply]
Found it where you said. https://github.com/Eugeny/tabby/blob/35f0d6908a0507ddb3a8540641d93c0ebbfc4fd6/tabby-terminal/src/api/baseTerminalTab.component.ts#L408 Thanks!

**Comment by lawweiliang** [comment]
@CoconutMacaroon   
Please close the discussion if the question was answered. 
It would be helpful to make the discussion section clean. 

Click `mark as answered` and `close the discussion`. 😄 

![image](https://user-images.githubusercontent.com/34936929/228098333-222ac3d4-4c39-4c26-a2a8-52906647e00e.png)

**Reply by CoconutMacaroon** [reply]
Sorry. I've closed it

**Comment by and7ey** [comment]
Do I understand correctly that the percentage should be at the last line of output?
It doesn't capture the output of my python script:

```bash
🎬 [24/55] (43.6%) IMG_9530.MOV
   🗓️ event: unknown
   🎞 extracting frames...
   📸 frames: 3
   👤 recognizing faces...
   👥 people: []
   🤖 sending to Qwen3-VL...
   ⏳ generating...
   ✅ Qwen done
➡️   IMG_9530.MOV
✨   IMG_9530 New Name.MOV

🎬 [25/55] (45.5%) IMG_5493.MOV
   🎞 extracting frames...
   📸 frames: 9
   👤 recognizing faces...
   👥 people: []
   🤖 sending to Qwen3-VL...
   ⏳ generating...
   ✅ Qwen done
➡️   IMG_5493.MOV
✨   IMG_5493 New Name.MOV
```

**Reply by Eugeny** [reply]
Percentage has to the in the last PTY output chunk - Python flushes stdout after each `print`, but you can disable it and flush manually after each progress "step"

**Reply by and7ey** [reply]
Thanks, @Eugeny. But I still don't understand well how it works.

I've prepared the following sample python3 script, it shows progress for a few seconds, but then it disappers.
```
import random
import time

messages = [
    "Connecting to server...",
    "Loading data...",
    "Analyzing content...",
    "Processing request...",
    "Checking results...",
    "Generating report...",
    "Synchronizing...",
    "Optimizing resources...",
    "Running tasks...",
    "Finalizing operation..."
]

for i in range(1, 11):
    print(random.choice(messages))
    time.sleep(1)

    print(random.choice(messages))
    time.sleep(1)

    print(f"Progress: {i * 10}%")

    if i < 10:
        time.sleep(60)

print("Completed! 100%")
```
